src/transformStateWithClones.js

```diff
@@ -8,6 +8,27 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      newState = { ...newState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
```

Comment:
[DONT REPEAT YOURSELF] - Instead of assigning to 'newState' in every case, you can modify 'newState' directly and push its copy to 'stateHistory' at the end of the loop cycle, but outside of the `switch` block, after current action is processed.